### ISSUE TITLE
 run HFVAE for 20 Newsgroup

### ISSUE THREAD

**Issue by rana-alshaikh** [probtorch/probtorch#41]
Thanks for making the code available, but is it possible to explain to me how to run the code of the model in (Structured Disentangled Representations)if I have the bag-of-word for documents or how you run it for 20Newsgroupe.

Thanks

**Comment by hao-w** [comment]
Hi, I'ill take 20newsgroup as an example and walk you through the implementation and training process.

For the data preprocessing,

**Comment by hao-w** [comment]
Hi, what you need to do is just to replace the monte-carlo objective a VAE model (e.g. NVDM) with the HFVAE objective, on the high level.  

Now let's say we want to train HFVAE for 20newsgroup, using the architecture of NVDM. There are several main steps:

1. Data preprocessing: we preprocessd the raw data to get bag-of-word representation. I personally wouldn't recommend using the processed dataset released by the authors of the NVDM paper ([https://github.com/ysmiao/nvdm](https://github.com/ysmiao/nvdm)),  since certain cases are not handled properly, such as numerics and stemming. 

Instead, we downloaded the 20newsgroup and ran the preprocessing script called 'preprocess_data.py' from [https://github.com/dallascard/scholar](https://github.com/dallascard/scholar), where they used NVDM model as a baseline in their paper. Their preprocessing is more reasonable.


2. Now you just need to implement the nvdm model in probtorch. To be more specific, you need to code up an encoder 
```
class Encoder(nn.Module): 
     def __init__(self, vocabulary_size=VOCABULARY_SIZE, 
                                 num_hidden=NUM_HIDDEN, 
                                 num_latents=NUM_LATENTS)
           super(self.__class__, self).__init__()
           self.enc_hidden = nn.Sequential(
               nn.Linear(vocabulary_size, num_hidden),
               nn.ReLU())
        
            self.z_mean = nn.Sequential(
               nn.Linear(num_hidden, num_latents))
            self.z_log_std = nn.Sequential(
               nn.Linear(num_hidden, num_latents))

    @expand_inputs
    def forward(self, documents, num_samples=None):
        q = probtorch.Trace()
        hidden = self.enc_hidden(documents)
        z_mean = self.z_mean(hidden)
        z_std = torch.exp(self.z_log_std(hidden))
        q.normal(z_mean, z_std, name='z')

        return q
```
and a decoder
```
class Decoder(nn.Module):
    def __init__(self, vocabulary_size=VOCABULARY_SIZE,
                       num_hidden=NUM_HIDDEN,
                       num_latents=NUM_LATENTS,
                       num_samples=NUM_SAMPLES,
                       batch_size=BATCH_SIZE):
        super(self.__class__, self).__init__()

        self.dec_document = nn.Sequential(
                           nn.Linear(num_latents, vocabulary_size),
                           nn.Softmax(dim=-1))
        
        self.prior_mean = torch.zeros((batch_size, num_latents)).cuda()
        
        self.prior_cov = torch.eye((num_latents)).cuda() 
        
    def forward(self, documents, q=None, num_samples=None):
        p = probtorch.Trace()  
        z = p.multivariate_normal(loc=self.prior_mean, covariance_matrix=self.prior_cov, value=q['z'], name='z')
        documents_recon = self.dec_document(z)
        p.loss(cross_entropy, documents_recon, documents, name='documents')
        return p

def cross_entropy(x_recon, x):
    return - (torch.log(x_recon + 1e-8) * x).sum(-1)
```
and finally you call the HFVAE objective to train the model
```
def elbo(q, p):
    if NUM_SAMPLES is None:
        ave_elbo = probtorch.objectives.marginal.elbo(q, p, sample_dim=None, batch_dim=0, beta=BETA, bias=BIAS)
        return ave_elbo
    else:
        ave_elbo = probtorch.objectives.marginal.elbo(q, p, sample_dim=0, batch_dim=1, beta=BETA, bias=BIAS)
        return ave_elbo
```
For the rest of the implementation, please check the example folder for detail.

Please let me know if you have any question.

**Comment by rana-alshaikh** [comment]
Hi,
Thanks for your respond
In the code you provided above, the 'documents'  is it the Bag of words matrix for all the documents, right? because I already have the bag-of-word representations and I don't need to preprocess any data, what I need are the followings
How to extract the document representation, the latent variable representation, and the word representations which is the weights of the [nn.Linear(num_latents, vocabulary_size)]
     
        
I will really appreciate your help in this, and regarding running the code
I will run the code as follow
```

from torchvision import datasets, transforms
import os 
if not os.path.isdir(DATA_PATH):
    os.makedirs(DATA_PATH)
train_data = torch.utils.data.DataLoader(BoW text file,
                batch_size=NUM_BATCH, shuffle=True)
test_data = torch.utils.data.DataLoader(BoW text file,
                batch_size=NUM_BATCH, shuffle=True)
    
```
```

def cuda_tensors(obj): 
    for attr in dir(obj):
        value = getattr(obj, attr)
        if isinstance(value, torch.Tensor):
            setattr(obj, attr, value.cuda())

enc = Encoder()
dec = Decoder()
if CUDA:
    enc.cuda()
    dec.cuda()
    cuda_tensors(enc)
    cuda_tensors(dec)

optimizer =  torch.optim.Adam(list(enc.parameters())+list(dec.parameters()),
                              lr=LEARNING_RATE,
                              betas=(BETA1, 0.999))`

```

I will appreciate if you tell me how to change the following code to obtain the representations

```

def train(data, enc, dec, optimizer):
    epoch_elbo = 0.0
    enc.train()
    dec.train()
    N = 0
    for b, (images, labels) in enumerate(data):
        if images.size()[0] == NUM_BATCH:
            N += NUM_BATCH
            images = images.view(-1, NUM_PIXELS)
            if CUDA:
                images = images.cuda()
            optimizer.zero_grad()
            q = enc(images, num_samples=NUM_SAMPLES)
            p = dec(images, q, num_samples=NUM_SAMPLES)
            loss = -elbo(q, p)
            loss.backward()
            optimizer.step()
            if CUDA:
                loss = loss.cpu()
            epoch_elbo -= float(loss.item())
    return epoch_elbo / N

def test(data, enc, dec):
    enc.eval()
    dec.eval()
    epoch_elbo = 0.0
    N = 0
    for b, (images, labels) in enumerate(data):
        if images.size()[0] == NUM_BATCH:
            N += NUM_BATCH
            images = images.view(-1, NUM_PIXELS)
            if CUDA:
                images = images.cuda()
            q = enc(images, num_samples=NUM_SAMPLES)
            p = dec(images, q, num_samples=NUM_SAMPLES)
            batch_elbo = elbo(q, p)
            if CUDA:
                batch_elbo = batch_elbo.cpu()
            epoch_elbo += float(batch_elbo.item())
    return epoch_elbo / N
```


I really  appreciate your help and your time

**Comment by hao-w** [comment]
So the questions are how to get the reconstructed documents, the latent variables, and the word representation. All of them are pretty straightforward once you are familiar with the Class **Trace** in probtorch. A **trace** is a dictionary-like container for stochastic nodes (variables). Entries in a **trace** are ordered and can be retrieved by the key. For instance, you can retrieve the following values as 

1) **reconstruction**
When the decoder is defined, we instantiate a trace variable **p** and store the reconstruction in it, indexed by the string "documents"
```
p.loss(cross_entropy, documents_recon, documents, name='documents')
```
Thus, when you pass a batch of documents to the VAE, you can retrieve that by the command
```
p['ducments'].value     ## SAMPLE_DIM * BATCH_DIM * VOCABULARY_SIZE
```
which are samples of the reconstructed documents for this batch.

2) **latent variables**
Similarly, the latent variables are stored in trace q, indexed by the name 'z'. So you can retrieve the latent variables by
```
q['z'].value
```
which are samples of latent variables of this batch. Alternatively, you can also check the learned paramters of the variational distribution q_\phi(latent | doc) by
```
q['z'].dist.loc
q['z'].dist.scale
``` 
which give you the mean and standard deviation of that normal distribution.

3) **word distribution**
In the NVDM case, the word distritbution really means the weights of the decoder, so the question is how to extract the weights of that linear layer, i.e. it is a pytorch operation
```
list(decoder.parameters())[0]
```
which gives you the weights of the first (and the only) linear layer.

Hope this helps.

**Comment by rana-alshaikh** [comment]
Thanks for your help
But I am still a little bit confused about where to pass the BoW matrix and in what form.  Is this correct because it gives me an error
'''
from torchvision import datasets, transforms
import os 
if not os.path.isdir(DATA_PATH):
    os.makedirs(DATA_PATH)
train_data = torch.utils.data.DataLoader(BoW text file,
                batch_size=NUM_BATCH, shuffle=True)
test_data = torch.utils.data.DataLoader(BoW text file,
                batch_size=NUM_BATCH, shuffle=True)
'''

**Comment by hao-w** [comment]
can you post the error message?

**Comment by rana-alshaikh** [comment]
I think I knew, it is about the format of the file you use JSON, and I use a text file 
the kernel just died when I run the code 
I don't want to waste your time I will try to solve it and put everything together and see if it works or not and then back to you :)

 thanks again

**Comment by alicanb** [comment]
Hi, I'm gonna close this issue due to inactivity. Feel free to reopen it if you still have questions.
